File: examples/python/linux-node-overview/src/host.py
from grafana_foundation_sdk.builders import common, dashboard, stat
from grafana_foundation_sdk.models.common import BigValueColorMode
from grafana_foundation_sdk.models.dashboard import ThresholdsMode, Threshold
from grafana_foundation_sdk.models import units

from .common import prometheus_query, uname_stat, default_stat


def uptime_stat() -> stat.Panel:
    return (
        default_stat()
        .title("Uptime")
        .description(
            "The duration of time that has passed since the last reboot or system start."
        )
        .unit(units.DurationSeconds)
        .with_target(
            prometheus_query(
                query='time() - node_boot_time_seconds{job=~"integrations/(node_exporter|unix)",cluster=~"$cluster",job=~"$job",instance=~"$instance"}',
            )
        )
        .thresholds(
            dashboard.ThresholdsConfig()
            .mode(ThresholdsMode.ABSOLUTE)
            .steps(
                [
                    Threshold(color="orange"),
                    Threshold(value=600, color="text"),
                ]
            )
        )
    )


def os_stat() -> stat.Panel:
    return (
        default_stat()
        .title("OS")
        .description("Operating system.")
        .with_target(
            prometheus_query(
                query='node_os_info{job=~"integrations/(node_exporter|unix)",cluster=~"$cluster",job=~"$job",instance=~"$instance"}',
            )
        )
        .reduce_options(
            common.ReduceDataOptions().calcs(["lastNotNull"]).fields("/^pretty_name$/")
        )
        .color_mode(BigValueColorMode.NONE)
    )


def hostname_stat() -> stat.Panel:
    return uname_stat(
        title="Hostname",
        description="System's hostname.",
        field="nodename",
    )


def kernel_version_stat() -> stat.Panel:
    return uname_stat(
        title="Kernel version",
        description="Kernel version of linux host.",
        field="release",
    )

File: examples/typescript/linux-node-overview/src/host.ts
import * as common from '@grafana/grafana-foundation-sdk/common';
import * as dashboard from '@grafana/grafana-foundation-sdk/dashboard';
import { PanelBuilder as StatBuilder } from '@grafana/grafana-foundation-sdk/stat';
import * as units from '@grafana/grafana-foundation-sdk/units';
import { defaultStat, prometheusQuery, tablePrometheusQuery, unameStat } from "./common";

export const uptimeStat = (): StatBuilder => {
  return defaultStat()
    .title("Uptime")
    .description("The duration of time that has passed since the last reboot or system start.")
    .withTarget(
      prometheusQuery('time() - node_boot_time_seconds{job=~"integrations/(node_exporter|unix)",cluster=~"$cluster",job=~"$job",instance=~"$instance"}', '')
    )
    .unit(units.DurationSeconds)
    .thresholds(
      new dashboard.ThresholdsConfigBuilder()
        .mode(dashboard.ThresholdsMode.Absolute)
        .steps([
          { value: null, color: "orange" },
          { value: 600, color: "text" },
        ])
    );
};

export const hostnameStat = (): StatBuilder => {
  return unameStat(
    "Hostname",
    "System's hostname.",
    "nodename",
  );
};

export const kernelVersionStat = (): StatBuilder => {
  return unameStat(
    "Kernel version",
    "Kernel version of linux host.",
    "release",
  );
};

export const osStat = (): StatBuilder => {
  return defaultStat()
    .title("OS")
    .description("Operating system.")
    .withTarget(
      tablePrometheusQuery('node_os_info{job=~"integrations/(node_exporter|unix)",cluster=~"$cluster",job=~"$job",instance=~"$instance"}', 'A')
    )
    .reduceOptions(
      new common.ReduceDataOptionsBuilder()
        .calcs(["lastNotNull"])
        .fields("/^pretty_name$/")
    )
    .colorMode(common.BigValueColorMode.None);
};
